Observe timestamp for uncertainty on every request

tbg opened the pull request:
Still got to write a test, but there's some awkward mocking involved.
## 

When executing on a Store, a transactional command now always returns a
timestamp off the local HLC clock. Thus, the first request to a node will
likely shrink the uncertainty interval for that node from `[x, x+250ms)`
to `[x, time_on_node)`, which should remove a lot of uncertainty-related
restarts.
One common occurrence of such restarts was spurious: When pushing an intent,
that intent will be encountered in the near future when retrying the write.
To avoid a restart in that case, we
- generally push not the to the minimal timestamp ahead of us but all the way
  up to the local HLC timestamp, and
- lower our uncertainty interval to this timestamp.

Added an originating Node's ID to roachpb.Error. Together with always observing
a timestamp, this allowed completely removing the active payload from
ReadWithinUncertaintyIntervalError and removed some questionable logic in
`replica_command.go`.

Fixes #2861.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4814)

<!-- Reviewable:end -->

spencerkimball commented:
Meta comment: Why not always return the timestamp off the local HLC clock (not just for transactional commands), and in addition use it to update the dist sender node's HLC clock. This would allow us to choose new timestamps which are less likely to require uncertainty restarts.

spencerkimball commented on kv/txn_coord_sender.go:
```diff
@@ -713,6 +705,17 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
+	// waiting until EndTransaction).
+	if pErr == nil && newTxn.Isolation == roachpb.SERIALIZABLE &&
+		!newTxn.OrigTimestamp.Equal(newTxn.Timestamp) {
+		log.Warningf("doing it %s vs %s", newTxn.OrigTimestamp, newTxn.Timestamp)
```
Do you want to keep this log statement?

tschottdorf replied:
Oops. Removed.

On Thu, Mar 3, 2016 at 10:38 AM Spencer Kimball <notifications@github.com>
wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/4814#discussion_r54896885>:
>
> > @@ -713,6 +705,17 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
> >  	} else {
> >  		newTxn.Update(pErr.GetTxn())
> >  	}
> > +
> > +	// If the request was successful but we're in a transaction which needs to
> > +	// restart but doesn't know it yet, let it restart now (as opposed to
> > +	// waiting until EndTransaction).
> > +	if pErr == nil && newTxn.Isolation == roachpb.SERIALIZABLE &&
> > +		!newTxn.OrigTimestamp.Equal(newTxn.Timestamp) {
> > +		log.Warningf("doing it %s vs %s", newTxn.OrigTimestamp, newTxn.Timestamp)
>
> Do you want to keep this log statement?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4814/files#r54896885>.
>

spencerkimball commented on kv/txn_coord_sender.go:
```diff
@@ -729,19 +732,10 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
-		newTxn.Timestamp.Forward(candidateTS)
+		// interval, we simply move past the last observed timestamp from the
+		// node.
+		newTxn.Timestamp.Forward(newTxn.GetObservedTimestamp(pErr.OriginNode).Add(0, 1))
```
Use `Timestamp.Next()` here.

tschottdorf replied:
Actually don't need .Next() at all, removed. Always a bit sketchy to
synthesize timestamps.

On Thu, Mar 3, 2016 at 10:39 AM Spencer Kimball <notifications@github.com>
wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/4814#discussion_r54897069>:
>
> >  		// If the reader encountered a newer write within the uncertainty
> > -		// interval, move the timestamp forward, just past that write or
> > -		// up to ObservedTimestamp, whichever comes first.
> > -		candidateTS := newTxn.ObservedTimestamp
> > -		candidateTS.Backward(t.ExistingTimestamp.Add(0, 1))
> > -		newTxn.Timestamp.Forward(candidateTS)
> > +		// interval, we simply move past the last observed timestamp from the
> > +		// node.
> > +		newTxn.Timestamp.Forward(newTxn.GetObservedTimestamp(pErr.OriginNode).Add(0, 1))
>
> Use Timestamp.Next() here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4814/files#r54897069>.
>

spencerkimball commented on kv/txn_coord_sender.go:
```diff
@@ -729,19 +732,10 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
-		candidateTS := newTxn.ObservedTimestamp
-		candidateTS.Backward(t.ExistingTimestamp.Add(0, 1))
-		newTxn.Timestamp.Forward(candidateTS)
+		// interval, we simply move past the last observed timestamp from the
```
s/we simply move past the last observed timestamp from the node./advance the txn's timestamp just past the last observed timestamp from the node./

tschottdorf replied:
done.

On Thu, Mar 3, 2016 at 10:42 AM Spencer Kimball <notifications@github.com>
wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/4814#discussion_r54897413>:
>
> >  		// If the reader encountered a newer write within the uncertainty
> > -		// interval, move the timestamp forward, just past that write or
> > -		// up to ObservedTimestamp, whichever comes first.
> > -		candidateTS := newTxn.ObservedTimestamp
> > -		candidateTS.Backward(t.ExistingTimestamp.Add(0, 1))
> > -		newTxn.Timestamp.Forward(candidateTS)
> > +		// interval, we simply move past the last observed timestamp from the
>
> s/we simply move past the last observed timestamp from the node./advance
> the txn's timestamp just past the last observed timestamp from the node./
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4814/files#r54897413>.
>

spencerkimball commented on roachpb/data.go:
```diff
@@ -809,12 +804,18 @@ func (t Transaction) Short() string {
 // operations in the transaction. When multiple calls are made for a single
 // nodeID, the lowest timestamp prevails. Calls with timestamps above
-// t.ObservedTimestamp have no effect.
+// t.ObservedTimestamp have no effect, unless t.ObservedTimestamp is zero.
```
Btw, I think it's quite confusing to call this variable `ObservedTimestamp`. This is the maximum ceiling for uncertainty interval. It has much less to do with observations. Could we rename this `UncertaintyCeilingTS` or back to `MaxTimestamp`?

spencerkimball commented on storage/store.go:
```diff
@@ -1412,7 +1412,45 @@ func (s *Store) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.Bat
+		// Reduce uncertainty restarts by attaching the above timestamp to the
+		// returned response or error (if in a txn). The caller can use it to
+		// shorten its uncertainty interval when it comes back to this node.
+		defer func() {
```
I think this will get a lot cleaner if you always return `now` in the response header. Then the caller can update its clock and possibly the observed timestamp corresponding to this node if the batch was part of a transaction.

tschottdorf replied:
I tried this initially and backed off of it, but giving it another shot.

spencerkimball commented on storage/store.go:
```diff
@@ -1465,31 +1502,30 @@ func (s *Store) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.Bat
-				if pErr.Index != nil {
-					pErr.SetErrorIndex(index.Index)
+			args := ba.Requests[index.Index].GetInner()
+			h := ba.Header // for the copy
```
s/for the copy/make a copy of the header for intent push/

tschottdorf replied:
done.

On Thu, Mar 3, 2016 at 10:54 AM Spencer Kimball <notifications@github.com>
wrote:

> In storage/store.go
> <https://github.com/cockroachdb/cockroach/pull/4814#discussion_r54899358>:
>
> > -				resolveIntents, pErr = s.resolveWriteIntentError(ctx, wiErr, rng, args, ba.Header, pushType)
> > -				if len(resolveIntents) > 0 {
> > -					if resErr := rng.resolveIntents(ctx, resolveIntents, false /* !wait */, true /* poison */); resErr != nil {
> > -						// When resolving asynchronously, errors should not
> > -						// usually be returned here, although there are some cases
> > -						// when they may be (especially when a test cluster is in
> > -						// the process of shutting down).
> > -						log.Warningf("asynchronous resolveIntents failed: %s", resErr)
> > -					}
> > -				}
> > -				// Make sure that if an index is carried in the error, it
> > -				// remains the one corresponding to the batch here.
> > -				if pErr.Index != nil {
> > -					pErr.SetErrorIndex(index.Index)
> > +			args := ba.Requests[index.Index].GetInner()
> > +			h := ba.Header // for the copy
>
> s/for the copy/make a copy of the header for intent push/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4814/files#r54899358>.
>

spencerkimball commented:
LGTM

tschottdorf commented:
> Meta comment: Why not always return the timestamp off the local HLC clock (not just for transactional commands), and in addition use it to update the dist sender node's HLC clock. This would allow us to choose new timestamps which are less likely to require uncertainty restarts.

Yep, that's #3525. Should be easy to do with your other suggestion which I'll give another go shortly.

tamird commented:
Reviewed 15 of 19 files at r1, 2 of 3 files at r2.
Review status: 17 of 19 files reviewed at latest revision, 10 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 515 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KByCJ33FXo6QcoSdGHS:-KByCJ33FXo6QcoSdGHT:1309594558) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d20e6ae18ec620515b29cd7cf2c15f6b1324196/kv/txn_coord_sender_test.go#L515)):*
why does this need to be a function?

---

*[roachpb/errors.proto, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KBy8lfEU8B4A0LpDhmQ:-KBy8lfEU8B4A0LpDhmR:1003438547) ([raw file](https://github.com/cockroachdb/cockroach/blob/715bfd1e91de157bde7af6e3363d911c525ad7db/roachpb/errors.proto#L62)):*
i think the word "future" is redundant here

---

*[roachpb/errors.proto, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KBy8umZDDjl7yf04p9t:-KBy8umZDDjl7yf04p9u:1751419181) ([raw file](https://github.com/cockroachdb/cockroach/blob/715bfd1e91de157bde7af6e3363d911c525ad7db/roachpb/errors.proto#L64)):*
s/a lower bound for a new timestamp which is admissible/the minimum timestamp which is outside the reader's uncertainty interval/

---

*[roachpb/errors.proto, line 250 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KBy95mT5a4VfEvo0eh_:-KBy95mT5a4VfEvo0eha:845464285) ([raw file](https://github.com/cockroachdb/cockroach/blob/715bfd1e91de157bde7af6e3363d911c525ad7db/roachpb/errors.proto#L250)):*
fantastic. can you update the comment on this guy while you're here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 17 of 19 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender_test.go, line 515 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KByCJ33FXo6QcoSdGHS:-KByX9pHxhfRbS-dV--v:-1157291696) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d20e6ae18ec620515b29cd7cf2c15f6b1324196/kv/txn_coord_sender_test.go#L515)):*
Because I'd otherwise have to hand-craft an `ObservedTimestamps` map.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 17 of 19 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender_test.go, line 515 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KByCJ33FXo6QcoSdGHS:-KByXeoFOqXZT_i8TTKM:-599819328) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d20e6ae18ec620515b29cd7cf2c15f6b1324196/kv/txn_coord_sender_test.go#L515)):*
Oh, it's used in expression position. Didn't see that from the diff.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@spencerkimball I've got some WIP for #3525, but there's some refactoring for clarity which is hard to pass on but is mostly orthogonal. Going to merge this and whip up the next PR for #3525 instead.

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: 17 of 19 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PTAL at the second commit.

tamird commented:
Reviewed 14 of 15 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1N7LGU9JRnshJxnRA:-KC1N7LGU9JRnshJxnRB:-1214767007) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/kv/txn_coord_sender.go#L715)):*
is this necessary?

---

*[kv/txn_coord_sender.go, line 718 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1NAwDASrau9Eni9uF:-KC1NAwDASrau9Eni9uG:1385623424) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/kv/txn_coord_sender.go#L718)):*
drive-by: this comment appears to be outdated

---

*[kv/txn_test.go, line 161 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1PNE5b8JWZgwW50GJ:-KC1PNE5b8JWZgwW50GK:-290498195) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/kv/txn_test.go#L161)):*
i dont think you need the extra `func` here

---

*[kv/txn_test.go, line 180 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1PmVirMuUgRj4wzDg:-KC1PmVirMuUgRj4wzDh:-1577904762) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/kv/txn_test.go#L180)):*
can you separate these error conditions? if txn.Proto.Epoch > 1 this error will be unintelligible.

---

*[kv/txn_test.go, line 185 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1QCgtYLC-QgnZR5AI:-KC1QCgtYLC-QgnZR5AJ:-1315072104) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/kv/txn_test.go#L185)):*
i think you can return `pErr` regardless

---

*[roachpb/data.proto, line 254 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1MXCh58q3LVlDzFg9:-KC1MXCh58q3LVlDzFgA:-1532890629) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/roachpb/data.proto#L254)):*
metacomment: it might be good to avoid referring to the actual field name in this comment to avoid having to rewrite the comment each time the field name changes.

---

*[roachpb/data.proto, line 255 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1MhXM9celbhDt6mpB:-KC1MhXM9celbhDt6mpC:351586957) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/roachpb/data.proto#L255)):*
certain_nodes is no longer a thing, right? again, perhaps avoid the specific field name.

---

*[storage/store.go, line 1429 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1O8ZRf_VyYbDN5i_L:-KC1O5wmkK-nSZO3F7QI:-1227437966) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store.go#L1429)):*
the comment above says that that `now` is attached to the error only if in a txn, but this seems to do it unconditionally.

---

*[storage/store.go, line 1435 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1OHs-XWthhA1Yf2fc:-KC1OHs-XWthhA1Yf2fd:1031005887) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store.go#L1435)):*
ditto re "if in a txn"

---

*[storage/store.go, line 1441 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1OV30zjVnpYWD32B6:-KC1OV30zjVnpYWD32B7:-2147444467) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store.go#L1441)):*
"if transactional" - where's the branch?

---

*[storage/store.go, line 1444 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1OQ3PMGL1IFqzFbwH:-KC1OQ3PMGL1IFqzFbwI:-917865187) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store.go#L1444)):*
s/observed/something else/g

---

*[storage/store_test.go, line 584 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RRFRmfEGhwcFkUY2:-KC1RRFRmfEGhwcFkUY3:376136431) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store_test.go#L584)):*
nit: newline after the opening brace and before the closing brace

---

*[storage/store_test.go, line 589 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RJ80EUvrXg5gravW:-KC1RJ81J_dEkqCCtz6_:1772290801) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store_test.go#L589)):*
please separate these error conditions. if the wallTime is wrong, this error will be opaque.

---

*[storage/store_test.go, line 608 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1Rbq6bd9Y821ETtiP:-KC1Rbq6bd9Y821ETtiQ:-1809756962) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store_test.go#L608)):*
log `exp` as well

---

*[storage/store_test.go, line 615 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RhCW3khCA6B-K7LJ:-KC1RhCW3khCA6B-K7LK:-14418809) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/storage/store_test.go#L615)):*
is it possible to reuse the store? seems sucky to have to set up the test each time.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
LGTM

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[roachpb/errors.proto, line 250 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1oDZw_pamIg3hW16X:-KC1oDZw_pamIg3hW16Y:-1316525059) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0abb3a3c191c2fd977b7c994ad59247fedeed/roachpb/errors.proto#L250)):*
It's not to be accessed directly in go code, but if this proto were used from other languages then there would be no alternative. That's why this used the proto name `txn` with a go `customname` of `UnexposedTxn`. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-:-KC1qGbXTrXXG-0aDumE:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 19 of 28 files reviewed at latest revision, 22 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1N7LGU9JRnshJxnRA:-KC1d7PgVv9deGEFFIpb:859631766) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/kv/txn_coord_sender.go#L715)):*
no, just appears in good good taste.

---

*[kv/txn_coord_sender.go, line 718 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1NAwDASrau9Eni9uF:-KC1dFkHAswyFA-qE0_I:-1073452576) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/kv/txn_coord_sender.go#L718)):*
We're still updating the txn, I'm moving it further down to where that happens.

---

*[roachpb/data.proto, line 255 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1MhXM9celbhDt6mpB:-KC1f_gRfOrvYUHumI9o:954269638) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/roachpb/data.proto#L255)):*
Maybe, but I think the clarity (or whatever amount of clarity there is) of the comments would suffer. Should be fine since we won't be able to go on renaming proto fields wildly very soon anyway.

---

*[roachpb/errors.proto, line 250 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1oDZw_pamIg3hW16X:-KC1r3mxxnuI0CeHvxzg:-133154104) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/roachpb/errors.proto#L250)):*
But even in other languages, the transaction should not be manipulated directly.

---

*[storage/store.go, line 1429 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1O8ZRf_VyYbDN5i_L:-KC1gC4V4zvgKuW4EZbl:631446628) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store.go#L1429)):*
The defer is guarded by `if ba.Txn != nil`.

---

*[storage/store.go, line 1441 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1OV30zjVnpYWD32B6:-KC1gJs9ncH6UzTJUfZY:241573666) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store.go#L1441)):*
this is inside `if ba.Txn != nil`. Updated comment.

---

*[storage/store_test.go, line 584 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RRFRmfEGhwcFkUY2:-KC1gtavqbQcByXe75oF:66751842) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store_test.go#L584)):*
acknored.

---

*[storage/store_test.go, line 615 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RhCW3khCA6B-K7LJ:-KC1hfwWPMM8m81M8_Zo:1386970498) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store_test.go#L615)):*
Possible, but I avoid it. Tends to require more bookkeeping (extra keys, ...). Test takes .01s, btw. Stores are cheap.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 19 of 28 files reviewed at latest revision, 22 unresolved discussions.

---

*[roachpb/errors.proto, line 250 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1oDZw_pamIg3hW16X:-KC2DJZrekaK4ssLbZ5M:153406596) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/roachpb/errors.proto#L250)):*
ah, stale comment. Anyway, I talked to @bdarnell who pointed out that other languages would only ever need to read this, so they wouldn't have that issue. We don't plan for kv client drivers in other languages, so going to leave as is for simplicity.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 9 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1N7LGU9JRnshJxnRA:-KC2FM6VvZkuMAOnqjeB:590729574) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/kv/txn_coord_sender.go#L715)):*
don't agree, but ok

---

*[roachpb/data.proto, line 255 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1MhXM9celbhDt6mpB:-KC2DfdEF1j7SkxPFSo_:-1170372503) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/roachpb/data.proto#L255)):*
heh, actually there's nothing stopping us from renaming proto fields - it's only the tags that need to remain constant.

either way, please remove the reference to "certain_nodes"

---

*[storage/store.go, line 1429 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1O8ZRf_VyYbDN5i_L:-KC2E3f53AlUHc9h3uVF:-1724504953) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store.go#L1429)):*
Ah, so it is. Perhaps the comment would be clearer if it said "we're in a txn, so ...."

---

*[storage/store_test.go, line 584 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RRFRmfEGhwcFkUY2:-KC2ElExZzgnxEZrYAW-:1025575645) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store_test.go#L584)):*
cc @bdarnell @petermattis for style guide.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[storage/store_test.go, line 584 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC1RRFRmfEGhwcFkUY2:-KC2GMv9M4YXTPCqTSQb:1438935692) ([raw file](https://github.com/cockroachdb/cockroach/blob/1542bfcbc53b5492c09dfdb7fa7389e43ae28f0f/storage/store_test.go#L584)):*
I would leave this to the author's preference.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->

tbg committed:
Observe timestamp for uncertainty on every request

When executing on a Store, a transactional command now always returns a
timestamp off the local HLC clock. Thus, the first request to a node will
likely shrink the uncertainty interval for that node from `[x, x+250ms)`
to `[x, time_on_node)`, which should remove a lot of uncertainty-related
restarts.
One common occurrence of such restarts was spurious: When pushing an intent,
that intent will be encountered in the near future when retrying the write.
To avoid a restart in that case, we
* generally push not to the minimal timestamp ahead of us but all the way
  up to the local HLC timestamp, and
* lower our uncertainty interval to this timestamp.

Added an originating Node's ID to roachpb.Error. Together with always observing
a timestamp, this allowed completely removing the active payload from
ReadWithinUncertaintyIntervalError and removed some questionable logic in
`replica_command.go`.

Fixes #2861.

tamird commented:
LGTM

---

Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 727 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC2ZXoV4Y2pP9RCJF6r:-KC2ZXoV4Y2pP9RCJF6s:803597096) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb55b71504a5da0670d4fc5b7a727a7bf975c63/kv/txn_coord_sender_test.go#L727)):*
```
pErr := test.Error
pErr.SetTxn(&txn)
return nil, &pErr
```
?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-:-KC2_G1hqtunAmBpZ4Cr:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 727 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814#-KC2ZXoV4Y2pP9RCJF6r:-KC2aIBgGmSirrl6MZKw:-1160250666) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb55b71504a5da0670d4fc5b7a727a7bf975c63/kv/txn_coord_sender_test.go#L727)):*
had that at first, but slight OCD discomfort at seeing a `pErr` which wasn't `*roachpb.Error`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4814)*
<!-- Sent from Reviewable.io -->
